### PR TITLE
Glharper/speaker package

### PR DESCRIPTION
**Pull request by glharper**
All APIs and basic tests added for voice profile/speaker recognition.

### CODE REVIEW

File: speaker/cfunctions.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (glecaros) [inline speaker/cfunctions.go:1]: why is this file needed?

File: speaker/speaker_identification_model.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
+
+package speaker
+
+import (
+	// "fmt"
```
Comment (glecaros) [inline speaker/speaker_identification_model.go:7]: nit: remove this comment

File: speaker/speaker_verification_model.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
+
+package speaker
+
+import (
+	// "fmt"
```
Comment (glecaros) [inline speaker/speaker_verification_model.go:7]: nit: comment

File: speaker/voice_profile.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
+
+package speaker
+
+import (
+	// "fmt"
```
Comment (glecaros) [inline speaker/voice_profile.go:7]: nit: comment

File: speaker/voice_profile_client.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+// Close performs cleanup of resources.
+func (client VoiceProfileClient) Close() {
+	client.Properties.Close()
+	C.voice_profile_client_release_handle(client.handle)
+}
+
+type CreateProfileOutcome struct {
+	common.OperationOutcome
+
+	profile *VoiceProfile
```
Comment (glecaros) [inline speaker/voice_profile_client.go:63]: as this contains something "Closable" it probably needs a Close member function

File: speaker/voice_profile_client.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+				outcome <- VoiceProfileEnrollmentOutcome{Result: nil, OperationOutcome: common.OperationOutcome{err}}
+			} else {
+				outcome <- VoiceProfileEnrollmentOutcome{Result: newResult, OperationOutcome: common.OperationOutcome{nil}}
+			}
+		}
+	}()
+	return outcome
+}
+
+type GetAllProfilesOutcome struct {
```
Comment (glecaros) [inline speaker/voice_profile_client.go:201]: same as above...

File: speaker/speaker_identification_model.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+// #include <speechapi_c_speaker_recognition.h>
+import "C"
+
+// SpeakerIdentificationModel is the class that defines a identification model to be used in speaker identification scenarios.
+type SpeakerIdentificationModel struct {
+	handle     C.SPXHANDLE
+}
+
+// NewSpeakerIdentificationModelFromHandle creates a SpeakerIdentificationModel instance from a valid handle. This is for internal use only.
+func NewSpeakerIdentificationModelFromHandle(handle common.SPXHandle) (*SpeakerIdentificationModel, error) {
```
Comment (glecaros) [inline speaker/speaker_identification_model.go:21]: does this need to be public?

File: speaker/speaker_recognition_result.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// Close releases the underlying resources
+func (result SpeakerRecognitionResult) Close() {
+	result.Properties.Close()
+	C.recognizer_result_handle_release(result.handle)
+}
+
+// NewSpeakerRecognitionResultFromHandle creates a SpeakerRecognitionResult from a handle (for internal use)
+func NewSpeakerRecognitionResultFromHandle (handle common.SPXHandle) (*SpeakerRecognitionResult, error) {
```
Comment (glecaros) [inline speaker/speaker_recognition_result.go:51]: this needs to be public?

File: speaker/speaker_verification_model.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+// #include <speechapi_c_speaker_recognition.h>
+import "C"
+
+// SpeakerVerificationModel is the class that defines a verification model to be used in speaker verification scenarios.
+type SpeakerVerificationModel struct {
+	handle     C.SPXHANDLE
+}
+
+// NewSpeakerVerificationModelFromHandle creates a SpeakerVerificationModel instance from a valid handle. This is for internal use only.
+func NewSpeakerVerificationModelFromHandle(handle common.SPXHandle) (*SpeakerVerificationModel, error) {
```
Comment (glecaros) [inline speaker/speaker_verification_model.go:21]: same public question =)

File: speaker/voice_profile.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+// #include <speechapi_c_speaker_recognition.h>
+import "C"
+
+// VoiceProfile is the class that defines voice profiles used in speaker recognition scenarios.
+type VoiceProfile struct {
+	handle     C.SPXHANDLE
+}
+
+// NewVoiceProfileFromHandle creates a VoiceProfile instance from a valid handle. This is for internal use only.
+func NewVoiceProfileFromHandle(handle common.SPXHandle) (*VoiceProfile, error) {
```
Comment (glecaros) [inline speaker/voice_profile.go:22]: same public question

File: speaker/voice_profile_phrase_result.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// Close releases the underlying resources
+func (result VoiceProfilePhraseResult) Close() {
+	result.Properties.Close()
+	C.recognizer_result_handle_release(result.handle)
+}
+
+// NewVoiceProfilePhraseResultFromHandle creates a VoiceProfilePhraseResult from a handle (for internal use)
+func NewVoiceProfilePhraseResultFromHandle (handle common.SPXHandle) (*VoiceProfilePhraseResult, error) {
```
Comment (glecaros) [inline speaker/voice_profile_phrase_result.go:48]: same

File: speaker/voice_profile_result.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// Close releases the underlying resources
+func (result VoiceProfileResult) Close() {
+	result.Properties.Close()
+	C.recognizer_result_handle_release(result.handle)
+}
+
+// NewVoiceProfileResultFromHandle creates a VoiceProfileResult from a handle (for internal use)
+func NewVoiceProfileResultFromHandle (handle common.SPXHandle) (*VoiceProfileResult, error) {
```
Comment (glecaros) [inline speaker/voice_profile_result.go:44]: same

File: speaker/speaker_recognition_result.go (hunk 8 above)
Comment (glharper) [reply speaker/speaker_recognition_result.go:51]: used in speaker_recognizer

File: speaker/voice_profile.go (hunk 10 above)
Comment (glharper) [reply speaker/voice_profile.go:22]: used in voice_profile_client

File: speaker/voice_profile_enrollment_result.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// Close releases the underlying resources
+func (result VoiceProfileEnrollmentResult) Close() {
+	result.Properties.Close()
+	C.recognizer_result_handle_release(result.handle)
+}
+
+// NewVoiceProfileEnrollmentResultFromHandle creates a VoiceProfileEnrollmentResult from a handle (for internal use)
+func NewVoiceProfileEnrollmentResultFromHandle (handle common.SPXHandle) (*VoiceProfileEnrollmentResult, error) {
```
[the remark this answers is not in the document]
Comment (glharper) [reply speaker/voice_profile_enrollment_result.go:67]: used in voice_profile_client

File: speaker/voice_profile_phrase_result.go (hunk 11 above)
Comment (glharper) [reply speaker/voice_profile_phrase_result.go:48]: used in voice_profile_client.GetActivationPhrasesAsync

File: speaker/voice_profile_result.go (hunk 12 above)
Comment (glharper) [reply speaker/voice_profile_result.go:44]: used in VoiceProfileClient.ResetProfileAsync and DeleteProfileAsync

File: speaker/speaker_verification_model.go (hunk 9 above)
Comment (glharper) [reply speaker/speaker_verification_model.go]: made private

File: speaker/speaker_identification_model.go (hunk 7 above)
Comment (glharper) [reply speaker/speaker_identification_model.go]: made private

File: speaker/speaker_recognition_result.go (hunk 8 above)
Comment (glharper) [reply speaker/speaker_recognition_result.go:51]: Didn't realize private meant package level, not class level. Will fix all.

**Review by glecaros — APPROVED** [review]
